pyomo/contrib/mpc/examples/cstr/run_mhe.py

```diff
@@ -0,0 +1,364 @@
+import pyomo.contrib.mpc as mpc
+
+from pyomo.dae import ContinuousSet
+from pyomo.contrib.mpc.examples.cstr.run_mpc import get_steady_state_data
+from pyomo.contrib.mpc.interfaces.var_linker import DynamicVarLinker
+from pyomo.contrib.mpc.examples.cstr.model import create_instance
+from pyomo.contrib.mpc.modeling.mhe_constructor import (
+    construct_measurement_variables_constraints,
+    construct_disturbed_model_constraints,
+    activate_disturbed_constraints_based_on_original_constraints,
+    get_cost_from_error_variables,
+)
+from pyomo.contrib.mpc.modeling.cost_expressions import (
+    get_tracking_cost_from_time_varying_setpoint,
+)
+
+
+def get_control_inputs(sample_time=2.0):
+    n_samples = 5
+    control_input_time = [i*sample_time for i in range(n_samples)]
+    control_input_data = {
+        'flow_in[*]': [0.4 + 0.1*i for i in range(n_samples)]
+    }
+    series = mpc.TimeSeriesData(control_input_data, control_input_time)
+    # Note that if we want a json representation of this data, we
+    # can always call json.dump(fp, series.to_serializable()).
+    return series
+
+
+def run_cstr_mhe(
+    initial_data,
+    samples_per_estimator_horizon=5,
+    sample_time=2.0,
+    ntfe_per_sample_estimator=5,
+    ntfe_plant=5,
+    simulation_steps=5,
+    tee=False,
+):
+    estimator_horizon = sample_time * samples_per_estimator_horizon
+    ntfe = ntfe_per_sample_estimator * samples_per_estimator_horizon
+    m_estimator = create_instance(horizon=estimator_horizon, ntfe=ntfe)
+    estimator_interface = mpc.DynamicModelInterface(
+        m_estimator, m_estimator.time
+    )
+    t0_estimator = m_estimator.time.first()
+
+    m_plant = create_instance(horizon=sample_time, ntfe=ntfe_plant)
+    plant_interface = mpc.DynamicModelInterface(m_plant, m_plant.time)
+
+    # Sets initial conditions and initializes
+    estimator_interface.load_data(initial_data)
+    plant_interface.load_data(initial_data)
+
+    #
+    # Construct sample-point set for measurements and model disturbances
+    #
+    sample_points = [
+        t0_estimator +
+        sample_time*i for i in range(samples_per_estimator_horizon+1)
+    ]
+    m_estimator.sample_points = ContinuousSet(initialize=sample_points)
+
+    #
+    # Construct components for measurements and measurement errors
+    #
+    m_estimator.estimation_block = pyo.Block()
+    esti_blo = m_estimator.estimation_block
+
+    measured_variables = [
+        pyo.Reference(m_estimator.conc[:, "A"])
+    ]
+    measurement_info = construct_measurement_variables_constraints(
+        m_estimator.sample_points,
+        measured_variables,
+    )
+    esti_blo.measurement_set = measurement_info[0]
+    esti_blo.measurement_variables = measurement_info[1]
+    # Measurement variables should be fixed all the time
+    esti_blo.measurement_variables.fix()
+    esti_blo.measurement_error_variables = measurement_info[2]
+    esti_blo.measurement_constraints = measurement_info[3]
```

Comment:
Are the measurement error variables necessary? We could just create measurement variables (or mutable parameters), then penalize the difference between estimates and measurements using `get_tracking_cost_expressions_from_time_varying_setpoint` (which should have its name changed since it can be used without a setpoint).

Personally, I think it would be nice to add as few variables/constraints to the model as possible.